### PR TITLE
WIP: upgrade to SymbolicUtils w/ fast terms

### PR DESCRIPTION
**Pull request by YingboMa**
This PR also deprecates `.op` and `.args` on `Term`s.

Most of the functionalities are upgraded except Hessian sparsity. @shashi is looking into that.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Is this breaking? @YingboMa take a look at Pumas and make sure it stays fine.

**Comment by isaacsas** [comment]
We'll need some Catalyst updates too.

**Comment by YingboMa** [comment]
Yeah, it's a little breaking, unfortunately. I will test those when this PR and the SymbolicUtils PR is finalized.

**Comment by ChrisRackauckas** [comment]
Yeah so this probably qualifies for a breaking release.

File: src/utils.jl [hunk 1]
```diff
@@ -309,10 +309,10 @@ xˍtt(t)
 ```
 """
 function diff2term(O)
-    isa(O, Term) || return O
+    (isa(O, Symbolic) && !isa(O, Sym)) || return O
```
Comment (YingboMa) [inline src/utils.jl:312]:
```suggestion
    istree(O)) || return O
```

**Comment by ChrisRackauckas** [comment]
Yeah no worries. I am fine with having it a bit common in this stage in the repo, especially for a 100x performance gain. It just needs to be done right and take care of our downstream buddies

**Comment by YingboMa** [comment]
If we are going to do a breaking release anyway. Let's also sneak in https://github.com/SciML/ModelingToolkit.jl/issues/684 :P

File: src/systems/jumps/jumpsystem.jl [hunk 2]
```diff
@@ -75,7 +75,7 @@ function generate_rate_function(js, rate)
 end
 
 function generate_affect_function(js, affect, outputidxs)
-    build_function(affect, states(js),
+    bf = build_function(map(x->x isa Equation ? x.rhs : x , affect), states(js),
```
Comment (shashi) [inline src/systems/jumps/jumpsystem.jl:78]: @isaacsas this should be correct, am I right?

Comment (isaacsas) [reply src/systems/jumps/jumpsystem.jl:78]: Looks good. Thanks!

**Comment by YingboMa** [comment]
All tests pass locally for me now.

**Comment by shashi** [comment]
@isaacsas @ChrisRackauckas (ping anyone else using this package downstream)

This is a breaking change! Let me summarize what this means to ModelingToolkit users:

- `Term` is not the only expression type now, there is also `Add`, `Mul`, and `Pow`.
- If you were dispatching on `Term`, now use `istree(x)` to check if `x` is of any of these expression types. All of them support `operation` and `arguments` accessors, so:
- `x.op` and `x.args` are deprecated in favor of `operation(x)` and `arguments(x)`.
Use these sed commands to make make this change across all files:
  - `sed -i 's/\(\b[a-zA-Z0-9_]*\b\)\.args/arguments\(\1\)/g' **/*.jl`
  - `sed -i 's/\(\b[a-zA-Z0-9_]*\b\)\.op/operation\(\1\)/g' **/**/*.jl`
 Note that `expr.args` where `expr` is an `Expr` will also be rewritten by this change, so be careful of what you commit (use `git add -p` to select the right changes, commit, and then run `git reset --hard` -- or however you usually do this.).

Here is what changed (and a migration guide) in SymbolicUtils.jl https://github.com/JuliaSymbolics/SymbolicUtils.jl/pull/154#issuecomment-756897615 if you need to get into the finer details.

**Comment by ChrisRackauckas** [comment]
Thanks, makes sense. Let's make sure own downstream libraries all get PRs too.

**Comment by YingboMa** [comment]
I’d like to do some more breaking changes before we merge it, like changing
the BipartiteGraph struct and the derivative.

**Comment by ChrisRackauckas** [comment]
Sounds good.

**Comment by YingboMa** [comment]
It's the annoying string LaTeX test again... Can we just delete it to test for nonempty string?

**Comment by ChrisRackauckas** [comment]
what's going on there?

**Comment by YingboMa** [comment]
Oh, this is actually due to the exp change. I will fix it.

**Comment by YingboMa** [comment]
Fingers crossed. CI should pass in this commit.

**Pull request merged**